packages/kbn-build/src/utils/parallelize.test.js

```diff
@@ -0,0 +1,92 @@
+import { parallelizeBatches } from './parallelize';
+
+// As promises resolve async, we use this helper to wait for all promises that
+// have been resolved to complete (aka call `then`).
+const tick = () =>
+  new Promise(resolve => {
+    setTimeout(resolve, 0);
+  });
+
+test('parallelizes batches', async () => {
+  const foo = createPromiseWithResolve();
+  const bar = createPromiseWithResolve();
+  const baz = createPromiseWithResolve();
+
+  const batches = [[foo, bar], [baz]];
+  const parallelize = parallelizeBatches(batches, obj => {
+    obj.called = true;
+    return obj.promise;
+  });
+
+  const completed = [];
+  parallelize.then(() => {
+    completed.push('parallelizeBatches');
+  });
+  foo.promise.then(() => {
+    completed.push('foo');
+  });
+  bar.promise.then(() => {
+    completed.push('bar');
+  });
+  baz.promise.then(() => {
+    completed.push('baz');
+  });
+
+  expect(foo.called).toBe(true);
+  expect(bar.called).toBe(true);
+  expect(baz.called).toBe(false);
+  expect(completed).toEqual([]);
+
+  bar.resolve();
+  await tick();
+
+  expect(completed).toEqual(['bar']);
+  expect(baz.called).toBe(false);
+
+  foo.resolve();
+  await tick();
+
+  expect(completed).toEqual(['bar', 'foo']);
+  expect(baz.called).toBe(true);
+
+  baz.resolve();
+  await tick();
+
+  expect(completed).toEqual(['bar', 'foo', 'baz', 'parallelizeBatches']);
+});
+
+test('rejects if any promise rejects', async () => {
+  const foo = createPromiseWithResolve();
+  const bar = createPromiseWithResolve();
+  const baz = createPromiseWithResolve();
+
+  const batches = [[foo, bar], [baz]];
+  const parallelize = parallelizeBatches(batches, obj => obj.promise);
+
+  const completed = [];
+  const failed = [];
+  parallelize.then(
+    () => {
+      completed.push('parallelizeBatches');
+    },
+    () => {
+      failed.push('parallelizeBatches');
+    }
+  );
+
+  foo.reject();
+  await tick();
+
+  expect(completed).toEqual([]);
+  expect(failed).toEqual(['parallelizeBatches']);
+});
+
+function createPromiseWithResolve() {
```

Comment:
In the past we were calling `PromiseWithResolveAndReject` as `Deffered` (https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred), but those times have passed :)

Reply:
`createDeferredish` :partyparrot: (I wasn't sure what to call this thing, but agree that `createDeferred` could work. I can make that change.)

Reply:
Nah, don't bother, it's not a big deal.